examples/s_expression.rs

```diff
@@ -0,0 +1,331 @@
+//! In this example we build an [S-expression](https://en.wikipedia.org/wiki/S-expression)
+//! parser and tiny [lisp](https://en.wikipedia.org/wiki/Lisp_(programming_language)) interpreter.
+//! Lisp is a simple type of language made up of Atoms and Lists, forming easily parsable trees.
+extern crate nom;
+
+use nom::{
+  branch::alt,
+  bytes::complete::tag,
+  character::complete::{alpha1, char, digit1, multispace0, multispace1},
+  combinator::{cut, map, opt},
+  error::{context, ParseError, VerboseError},
+  multi::many0,
+  sequence::{delimited, preceded, terminated, tuple},
+  IResult,
+};
+
+/// We start by defining the types that define the shape of data that we want.
+/// In this case, we want something tree-like
+
+/// Starting from the most basic, we define some built-in functions that our lisp has
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum BuiltIn {
+  Plus,
+  Minus,
+  Times,
+  Divide,
+  Equal,
+  Not,
+}
+
+/// We now wrap this type and a few other primitives into our Atom type.
+/// Remember from before that Atoms form one half of our language.
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum Atom {
+  Num(i32),
+  Keyword(String),
+  Boolean(bool),
+  BuiltIn(BuiltIn),
+}
+
+/// The remaining half is Lists. We implement these as recursive Expressions.
+/// For a list of numbers, we have `'(1 2 3)`, which we'll parse to:
+/// ```
+/// Expr::Quote(vec![Expr::Constant(Atom::Num(1)),
+///                  Expr::Constant(Atom::Num(2)),
+///                  Expr::Constant(Atom::Num(3))])
+/// Quote takes an S-expression and prevents evaluation of it, making it a data
+/// structure that we can deal with programmatically. Thus any valid expression
+/// is also a valid data structure in Lisp itself.
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum Expr {
+  Constant(Atom),
+  /// (func-name arg1 arg2)
+  Application(Box<Expr>, Vec<Expr>),
+  /// (if predicate do-this)
+  If(Box<Expr>, Box<Expr>),
+  /// (if predicate do-this otherwise-do-this)
+  IfElse(Box<Expr>, Box<Expr>, Box<Expr>),
+  /// '(3 (if (+ 3 3) 4 5) 7)
+  Quote(Vec<Expr>),
+}
+
+/// Continuing the trend of starting from the simplest piece and building up,
+/// we start by creating a parser for the built-in functions.
+fn parse_builtin<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, BuiltIn, E> {
```

Comment:
you can avoid the generic `E` parameter, and directly bring the `nom::error::VerboseError` type, since only this one will be used (I made the JSOn example generic to demonstrate how various error types work but it's not necessary everywhere)